File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:

        ### RECURSIVE APPROACH ###
        # def rob(n,nums):
        #     if n==0: return nums[0]
        #     if n<=-1: return 0
        #     pick = nums[n]+rob(n-2,nums)
        #     notpick = rob(n-1,nums)
        #     return max(pick,notpick)
        # n = len(nums)
        # ans = rob(n-1,nums)
        # return ans



        #### Memoization #####
        # def f(index,nums,dp):
        #     if index == 0 : return nums[0]
        #     if index<0 : return 0
        #     if dp[index] != -1: return dp[index]
        #     pick = nums[index] + f(index-2,nums,dp)
        #     not_pick = f(index-1,nums,dp)
        #     dp[index] = max(pick,not_pick)
        #     return dp[index]
        # n = len(nums)
        # dp = [-1] * (len(nums))
        # ans = f(n-1,nums,dp)
        # return ans


        ### Tabulation ###
        # n = len(nums)
        # dp = [-1] * n
        # dp[0]  = nums[0]

        # for i in range(1,n):
        #     not_pick = dp[i-1]
        #     pick = nums[i]
        #     if i>1:
        #         pick+=dp[i-2]
        #     dp[i] = max(pick,not_pick)
        # return dp[n-1]



        ### Space Optimization ###

        n = len(nums)
        
        prev  = nums[0]
        prev2 = 0

        for i in range(1,n):
            not_pick = prev
            pick = nums[i]
            if i>1:
                pick+=prev2
            curr = max(pick,not_pick)
            prev2 = prev
            prev = curr
        return prev

File: 0198-house-robber/0198-house-robber.java
class Solution {
    public int rob(int[] nums) {
        int prev2 = 0;
        int prev1 = 0;
        int curr = 0;
        for(int i=nums.length-1;i>=0;i--){
            curr = Math.max(nums[i]+prev2,prev1);
            prev2 = prev1;
            prev1 = curr;
        }
        return curr;
    }
}



// class Solution {
//     public int rob(int[] nums) {
//         int [] dp = new int[nums.length+2];
//         for(int i=nums.length-1;i>=0;i--){
//             dp[i] = Math.max(nums[i]+dp[i+2],dp[i+1]);
//         }
//         return dp[0];
//     }
// }

// class Solution {
//     public int f(int[] nums,int index,int[] dp){
//         if(index>=nums.length) return 0;
//         if(dp[index]!=-1) return dp[index];
//         int pick = nums[index] + f(nums,index+2,dp);
//         int not_pick = 0 + f(nums,index+1,dp);
//         return dp[index] = Math.max(pick,not_pick);
//     }
//     public int rob(int[] nums) {
//         int [] dp = new int[nums.length];
//         Arrays.fill(dp,-1);
//         return f(nums,0,dp);
//     }
// }


// class Solution {
//     public int f(int[] nums,int index){
//         if(index>=nums.length) return 0;
//         int pick = nums[index] + f(nums,index+2,dp);
//         int not_pick = 0 + f(nums,index+1,dp);
//         return Math.max(pick,not_pick);
//     }
//     public int rob(int[] nums) {
//         return f(nums,0,dp);
//     }
// }